block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -879,6 +907,76 @@ object BlockDagFileStorage {
       }
       .toSet
 
+  private def calculateInvalidBlocksCrc[F[_]: Monad](
+      invalidBlocks: List[BlockMetadata]
+  ): Crc32[F] =
+    Crc32[F](
+      invalidBlocks
+        .foldLeft(ByteString.EMPTY) {
+          case (byteString, blockMetadata) =>
+            val blockBytes = blockMetadata.toByteString
+            byteString.concat(blockBytes.size().toByteString.concat(blockBytes))
+        }
+        .toByteArray
+    )
+
+  private def readInvalidBlocksLog[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F]
+  ): F[List[BlockMetadata]] = {
+    def readRec(
+        result: List[BlockMetadata]
+    ): F[List[BlockMetadata]] =
+      for {
+        blockSizeOpt <- randomAccessIO.readInt
+        result <- blockSizeOpt match {
+                   case Some(blockSize) =>
+                     val blockMetaBytes = Array.ofDim[Byte](blockSize)
+                     for {
+                       _             <- randomAccessIO.readFully(blockMetaBytes)
+                       blockMetadata <- Sync[F].delay { BlockMetadata.fromBytes(blockMetaBytes) }
+                       result        <- readRec(blockMetadata :: result)
+                     } yield result
+                   case None =>
+                     result.reverse.pure[F]
+                 }
+      } yield result
+    readRec(List.empty)
+  }
+
+  private def validateInvalidBlocks[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      readInvalidBlocksCrc: Long,
+      invalidBlocks: List[BlockMetadata]
+  ): F[(List[BlockMetadata], Crc32[F])] = {
+    val fullCalculatedCrc = calculateInvalidBlocksCrc[F](invalidBlocks)
+    fullCalculatedCrc.value.flatMap { fullCalculatedCrcValue =>
+      if (fullCalculatedCrcValue == readInvalidBlocksCrc) {
+        (invalidBlocks, fullCalculatedCrc).pure[F]
+      } else if (invalidBlocks.nonEmpty) {
+        val withoutLastCalculatedCrc =
+          calculateInvalidBlocksCrc[F](invalidBlocks.init)
+        withoutLastCalculatedCrc.value.flatMap { withoutLastCalculatedCrcValue =>
+          if (withoutLastCalculatedCrcValue == readInvalidBlocksCrc) {
+            val lastRecord           = invalidBlocks.last
+            val lastRecordSize: Long = 4L + lastRecord.toByteString.size
+            for {
+              length <- randomAccessIO.length
+              _      <- randomAccessIO.setLength(length - lastRecordSize)
+            } yield (invalidBlocks.init, withoutLastCalculatedCrc)
+          } else {
+            Sync[F].raiseError[(List[BlockMetadata], Crc32[F])](
+              InvalidBlocksIsCorrupted
+            )
+          }
+        }
+      } else {
+        Sync[F].raiseError[(List[BlockMetadata], Crc32[F])](
+          InvalidBlocksIsCorrupted
+        )
+      }
+    }
```

Comment:
would it be possible to make this easier to read?
e.g.:

- name the if conditions

- extract the second flatMap method